x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/support/mapper/NativeRoleMappingStore.java

```diff
@@ -384,30 +382,10 @@ public void onSecurityIndexStateChange(SecurityIndexManager.State previousState,
             || isIndexDeleted(previousState, currentState)
             || Objects.equals(previousState.indexUUID, currentState.indexUUID) == false
             || previousState.isIndexUpToDate != currentState.isIndexUpToDate) {
-            refreshRealms(ActionListener.noop(), null);
-        }
-    }
-
-    private <Result> void refreshRealms(ActionListener<Result> listener, Result result) {
-        if (realmsToRefresh.isEmpty()) {
-            listener.onResponse(result);
-            return;
+            // the notification that the index state changed is received on every node
+            // this means that here we need only to invalidate the local realm caches only
+            clearRealmCachesOnLocalNode();
```

Comment:
This is almost a bug fix.

Reply:
Just double checking my understanding: Previously, this receive a state change notification, and then every node would send out a message telling all nodes to invalidate their caches. Now, every nodes gets the state change and invalidates their caches locally. Correct?

Assuming I read that right, damn, nice job.